### PR TITLE
feat(clusters): health matters, veto if unhealthy, notify

### PR DESCRIPTION
**Pull request by emjburns**
closes https://github.com/spinnaker/keel/issues/1506
closes https://github.com/spinnaker/keel/issues/1401

(note: all timing is a guess! feel free to offer suggestions)

This PR introduces an "UnhealthyVeto" that will pause resource checks for 5 minutes to let a resource recover (if there is no diff).
**NOTE**:  this won't happen while a deploy is going, because we aren't checking the resource while a deploy is going

When we detect a cluster is unhealthy with no diff, we notify based on the environment config.
- we notify immediately
- we notify once a day while unhealthy
- we clear the notification if it's healthy

The notification support is flexible and uses the new bot. There are some assumptions that are documented (because I wanted to keep it simple at first) but we could easily make this more complicated.

### CODE REVIEW

File: keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/resource/ClusterHandler.kt [hunk 1]
```diff
@@ -873,6 +874,13 @@ class ClusterHandler(
     val healthy: Boolean = activeServerGroups.all {
       it.instanceCounts?.isHealthy(resource.spec.deployWith.health) == true
     }
+
```
Comment (emjburns) [inline keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/resource/ClusterHandler.kt:877]: This event, emitted from the cluster handler and titus cluster handler, propagates the health information to the UnhealthyVeto

File: keel-echo/src/main/kotlin/com/netflix/spinnaker/config/KeelNotificationConfig.kt [hunk 2]
```diff
@@ -2,7 +2,7 @@ package com.netflix.spinnaker.config
 
 import org.springframework.boot.context.properties.ConfigurationProperties
```
Comment (emjburns) [inline keel-echo/src/main/kotlin/com/netflix/spinnaker/config/KeelNotificationConfig.kt:4]: generalized this to use for all notifications

File: keel-echo/src/main/kotlin/com/netflix/spinnaker/keel/echo/ManualJudgementNotifier.kt [hunk 3]
```diff
@@ -18,13 +18,13 @@ import org.springframework.stereotype.Component
 
 @Component
 @Configuration
-@EnableConfigurationProperties(ManualJudgementNotificationConfig::class)
+@EnableConfigurationProperties(KeelNotificationConfig::class)
 /**
```
Comment (emjburns) [inline keel-echo/src/main/kotlin/com/netflix/spinnaker/keel/echo/ManualJudgementNotifier.kt:22]: I think we could move this into the new notifier if we wanted manual judgment notifications to repeat more than once. However, I don't think we want that right now, so I think it belongs in a separate Notifer.

File: keel-echo/src/main/kotlin/com/netflix/spinnaker/keel/echo/model/EchoNotification.kt [hunk 4]
```diff
@@ -13,8 +13,9 @@ data class EchoNotification(
   val templateGroup: String? = null,
   val severity: Severity,
   val source: Source? = null,
-  val interactiveActions: InteractiveActions?,
-  val additionalContext: Map<String, Any?>? = null
+  val interactiveActions: InteractiveActions? = null,
+  val additionalContext: Map<String, Any?>? = null,
+  val useInteractiveBot: Boolean = true
```
Comment (emjburns) [inline keel-echo/src/main/kotlin/com/netflix/spinnaker/keel/echo/model/EchoNotification.kt:18]: This tells echo to use the new sexy spin bot

File: keel-web/src/main/kotlin/com/netflix/spinnaker/keel/rest/AdminController.kt [hunk 5]
```diff
@@ -72,4 +74,13 @@ class AdminController(
   ) {
     adminService.backFillArtifactMetadata(type)
   }
+
```
Comment (emjburns) [inline keel-web/src/main/kotlin/com/netflix/spinnaker/keel/rest/AdminController.kt:77]: an admin endpoint for easily testing out the notification sending

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/veto/unhealthy/UnhealthyVeto.kt [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+      stack = spec.moniker.stack ?: "(none)",
+      detail = spec.moniker.detail ?: "(none)"
+    )
+    val resourceUrl = "$spinnakerBaseUrl/#/applications/${resource.application}/clusters?${params.toURL()}"
+    return NotifierMessage(
+      subject = "${resource.spec.displayName} is unhealthy",
+      body = "<$resourceUrl|${resource.id}> is unhealthy and there is not a diff. " +
+        "Spinnaker cannot correct this problem automatically. " +
+        "Please take manual action to fix it."
+    )
```
Comment (erikmunson) [inline keel-core/src/main/kotlin/com/netflix/spinnaker/keel/veto/unhealthy/UnhealthyVeto.kt:92]: few detail-oriented thoughts:

1. when we're sending this notification do we have awareness of how long things have been unhealthy? if so, I'm wondering if we could re-word it to something like "<resource> has been unhealthy for 2 minutes...[etc.]"
2. I think the definite tone here ("Spinnaker cannot correct", "Please take manual action") is reassuring, but may not always be completely accurate. Giving a sense of total confidence has the potential to send users into "action mode" when they may need to investigate more before deciding to take action. Is there a way to reword this to still be confidence-boosting without suggesting we know exactly what needs to be done? for example, here's the language we use for the unhappy status:
> Spinnaker has been trying to correct a difference, but taking action hasn't helped. Manual intervention might be required. You can click History to see more.

3. "and there is not a diff" is slightly technical to me — is that detail useful/actionable in this notification?

Comment (emjburns) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/veto/unhealthy/UnhealthyVeto.kt:92]: Let me try to address this feedback. I think having a link to the history would be super useful, and I think I can give some sense of timing. I think linking to the history could provide the "no diff" message (what I'm trying to convey is "we have your resource right where you want it, pal, and something is not right)

Comment (luispollo) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/veto/unhealthy/UnhealthyVeto.kt:92]: Agree with "there is not a diff" being too technical. I suspect a lot of users won't understand what that means. How about something like this: 
```suggestion
      subject = "${resource.spec.displayName} is unhealthy",
      body = "Spinnaker has detected that <$resourceUrl|${resource.id}> matches its desired state, but is currently unhealthy. " +
        "We cannot correct this problem automatically. Manual intervention might be required. " +
        "You can view the resource history for more information."
    )
```

Comment (emjburns) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/veto/unhealthy/UnhealthyVeto.kt:92]: Screen Shot 2020-09-30 at 5 12 23 PM

Updated to this, what do you two think?

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/events/ResourceHealthEvent.kt [hunk 7]
```diff
@@ -0,0 +1,10 @@
+package com.netflix.spinnaker.keel.events
+
+/**
+ * An event emitted when we calculate resource health
+ */
+data class ResourceHealthEvent(
+  val resourceId: String,
+  val application: String,
+  val healthy: Boolean = true
```
Comment (luispollo) [inline keel-core/src/main/kotlin/com/netflix/spinnaker/keel/events/ResourceHealthEvent.kt:9]: Is it worth adding a `comment` or `details` field (probably only useful for the unhealthy case)?

Comment (luispollo) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/events/ResourceHealthEvent.kt:9]: Oh, never mind, these won't be persisted in history.

**Review by luispollo — COMMENTED** [review]
This looks great overall, @emjburns! Thanks for working on it!
I had a few questions and a handful of comments, nothing major. Hope the review helps!

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/persistence/NotifierRepository.kt [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+   * Assumption: each notifier sends only one type of message
+   * @return true if we should notify right now
+   */
+  abstract fun addNotification(resourceId: String, notifier: Notifier): Boolean
+
+  /**
+   * Clears a notification from the list of ongoing notifications.
+   * Does nothing if notification does not exist.
+   */
+  abstract fun clearNotification(resourceId: String, notifier: Notifier)
```
[the remark this answers is not in the document]
Comment (luispollo) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/persistence/NotifierRepository.kt:27]: P.S. The way this would feel more natural to me would be something like this:

```kotlin
enum class NotificationType {
  UNHEALTHY_RESOURCE
}

enum class NotificationScope {
  RESOURCE,
  // ...
}

abstract class Notification(
  val type: NotificationType,
  val scope: NotificationScope,
  val entityId: String, // resource ID, environment ID, etc.
  val subject: String,
  val message: String
)

data class ResourceUnhealthyNotification(
  val resource: Resource<*>
) : Notification(NotificationType.RESOURCE_UNHEALTHY, NotifiationScope.RESOURCE, resource.id, "Unhealthy resource", "Resource ${resource.id} is unhealthy")
```

Then your repository methods would look like:
```kotlin
abstract fun addNotification(notification: Notification)
```

So, actually no need for generics, just a bit of inheritance.

**Review by lorin — COMMENTED** [review]
Haven't had a chance to finish reviewing yet, but wanted to post my comments so far. 

My main question, which was inline, was about what happens in the scenario when a cluster goes unhealthy at the same time as a diff occurred.

**Comment by lorin** [comment]
One thing that wasn't clear to me: It looks like the `UnhealthyVeto` will only ever veto a resource if there's no diff. But I thought that the purpose of a veto is to prevent actuations, and actuations only happen if there's a diff!

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/persistence/NotifierRepository.kt [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+  open val clock: Clock,
+  @Value("notify.waiting-duration") var waitingDuration: String = "P1D"
+) {
+
+  /**
+   * Adds a notification to the list of ongoing notifications.
+   * Assumption: each notifier sends only one type of message
+   * @return true if we should notify right now
+   */
+  abstract fun addNotification(resourceId: String, notifier: Notifier): Boolean
```
[the remark this answers is not in the document]
Comment (emjburns) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/persistence/NotifierRepository.kt:21]: Good point. I'll add that so it's easy to do later!

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/persistence/NotifierRepository.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (emjburns) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/persistence/NotifierRepository.kt:27]: So we aren't saving the notification, we are just relying on the message to keep being sent via the event. This class is essentially throttling the message (it's saying, "we can't send another message until 1 day has passed, I'll throw this message out"). The other way to do things would be to save the notification in the database (like you suggested) and then iterate through the saved notifications and send them if it's time. I opted not to do this because there is no way to not send the event every time, so it seemed useless to save the message.

Basically, that means if you want to send two different types of "unhealthy" notifications for the same entity you'll need two different `Notifier`s, because we throttle based on entity id + notifier type (we do not take into account the message content at all).

Does that help?

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/persistence/UnhealthyVetoRepository.kt [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+  abstract fun isHealthy(
+    resourceId: String
+  ): Boolean
+
+  abstract fun markUnhealthy(
+    resourceId: String,
+    application: String
+  )
+
+  abstract fun markAllowed(
```
[the remark this answers is not in the document]
Comment (emjburns) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/persistence/UnhealthyVetoRepository.kt:32]: If the resource is unhealthy, we're still going to check it every 5 minutes. "Allowed time" (words are hard) represents the time we last allowed the resource to be checked instead of vetoing it. Marking it healthy isn't correct because the resource is still unhealthy

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/services/AdminService.kt [hunk 11]
```diff
@@ -112,4 +115,8 @@ class AdminService(
       }
     }
   }
+
+  fun sendNotificationEvent(event: ResourceNotificationEvent) {
```
[the remark this answers is not in the document]
Comment (emjburns) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/services/AdminService.kt:119]: It's used in the Admin controller, to be able to send notifications to test them out.

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/veto/unhealthy/ClusterViewParams.kt [hunk 12]
```diff
@@ -0,0 +1,17 @@
+package com.netflix.spinnaker.keel.veto.unhealthy
+
+/**
+ * Params for constructing a cluster view link
+ * If you want to search for stuff without a stack or detail,
+ * you must pass "none" for those
```
[the remark this answers is not in the document]
Comment (emjburns) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/veto/unhealthy/ClusterViewParams.kt:6]: "(none)", i'll update.

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/veto/unhealthy/UnhealthyVeto.kt [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+import org.springframework.context.ApplicationEventPublisher
+import org.springframework.context.event.EventListener
+import org.springframework.stereotype.Component
+import java.time.Clock
+import java.time.Duration
+import java.time.Instant
+
+/**
+ * A veto that stops keel from checking a resource while it's unhealthy,
+ * because there's nothing we can do to make it healthy
```
[the remark this answers is not in the document]
Comment (emjburns) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/veto/unhealthy/UnhealthyVeto.kt:28]: Do you feel it's important to end these with a period?

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/veto/unhealthy/UnhealthyVeto.kt [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+
+    return if (shouldAllow) {
+      log.debug("Resource $resourceId has been ignored for ${friendlyTime(ignoreDuration)}, allowing a recheck.")
+      unhealthyVetoRepository.markAllowed(resourceId)
+      allowedResponse()
+    } else {
+      // if we no longer have the resource marked as unhealthy, allow it to be checked
+      val timeNoticed = unhealthyVetoRepository.getNoticedTime(resource.id) ?: return allowedResponse()
+      publisher.publishEvent(ResourceNotificationEvent(resourceId, UNHEALTHY, message(resource, timeNoticed)))
+      deniedResponse(message = "Resource is unhealthy with no diff, waiting ${friendlyTime(ignoreDuration)} before rechecking.", vetoArtifact = false)
```
[the remark this answers is not in the document]
Comment (emjburns) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/veto/unhealthy/UnhealthyVeto.kt:64]: It will show up in the history view I believe (It's a little hard to test out locally, but that's my understanding)

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/veto/unhealthy/UnhealthyVeto.kt [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+      detail = spec.moniker.detail ?: "(none)"
+    )
+    val resourceUrl = "$spinnakerBaseUrl/#/applications/${resource.application}/clusters?${params.toURL()}"
+
+    val durationUnheathy = Duration.between(timeNoticed, clock.instant())
+
+    return NotifierMessage(
+      subject = "${resource.spec.displayName} is unhealthy",
+      body = "<$resourceUrl|${resource.id}> has been unhealthy for ${friendlyDuration(durationUnheathy)}. " +
+        "Spinnaker does not detect any changes to this resource. " +
```
[the remark this answers is not in the document]
Comment (emjburns) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/veto/unhealthy/UnhealthyVeto.kt:98]: maybe I can put "configuration changes" instead? is that clearer?

File: keel-echo/src/main/kotlin/com/netflix/spinnaker/keel/echo/ResourceNotifier.kt [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+      severity = EchoNotification.Severity.NORMAL,
+      source = EchoNotification.Source(
+        application = application
+      ),
+      additionalContext = mapOf(
+        "formatter" to "MARKDOWN",
+        "subject" to event.message.subject,
+        "body" to event.message.body
+      ),
+      interactiveActions = generateInteractiveConfig(event.resourceId, event.notifier.name)
```
[the remark this answers is not in the document]
Comment (emjburns) [reply keel-echo/src/main/kotlin/com/netflix/spinnaker/keel/echo/ResourceNotifier.kt:99]: It's defaulted on the echo config so you don't have to specify because we want all keel notifications to use that bot

File: keel-echo/src/main/kotlin/com/netflix/spinnaker/keel/echo/ResourceNotifier.kt [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+    )
+  }
+
+  private fun generateInteractiveConfig(resourceId: String, notifier: String): InteractiveActions? =
```
[the remark this answers is not in the document]
Comment (emjburns) [reply keel-echo/src/main/kotlin/com/netflix/spinnaker/keel/echo/ResourceNotifier.kt:103]: But you can specify the color in it! hehehe

File: keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlNotifierRepository.kt [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+  override fun addNotification(resourceId: String, notifier: Notifier): Boolean {
+    sqlRetry.withRetry(READ) {
+      jooq.select(NOTIFIER.NOTIFY_AT)
+        .from(NOTIFIER)
+        .where(NOTIFIER.RESOURCE_ID.eq(resourceId))
+        .and(NOTIFIER.NOTIFIER_.eq(notifier.name))
+        .fetchOne(NOTIFIER.NOTIFY_AT)
+    }?.let { notificationTime ->
+      // if record exists already, return whether or not to notify
+      return notificationTime < clock.millis()
```
[the remark this answers is not in the document]
Comment (emjburns) [reply keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlNotifierRepository.kt:27]: because it doesn't work nicely (locally, time zones are wacky and I tried for an hour to make it work) and I strongly prefer instant

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/veto/unhealthy/ClusterViewParams.kt [hunk 19]
```diff
@@ -0,0 +1,17 @@
...
+ */
+data class ClusterViewParams(
+  val acct: String,
+  val q: String,
```
[the remark this answers is not in the document]
Comment (emjburns) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/veto/unhealthy/ClusterViewParams.kt:10]: Honestly, I think query but not sure! This is for generating the UI deep link. I just stole what erik gave me

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/veto/unhealthy/UnhealthyVeto.kt [hunk 20]
```diff
@@ -0,0 +1,120 @@
...
+) : Veto {
+
+  private val log by lazy { LoggerFactory.getLogger(javaClass) }
+
+  override suspend fun check(resource: Resource<*>): VetoResponse {
+    val resourceId = resource.id
+    val hasDiff = diffFingerprintRepository.diffCount(resourceId) > 0
+
+    if (unhealthyVetoRepository.isHealthy(resourceId) || hasDiff) {
+      // we want keel to take action if there's a diff
```
[the remark this answers is not in the document]
Comment (emjburns) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/veto/unhealthy/UnhealthyVeto.kt:48]: Actually, that's not true. keel will detect that there's a diff in both configuration and enabled status and will launch a new task reverting whatever the user did while cloning. (we only disable the oldest if there is no other diff)

[the remark this answers is not in the document]
Comment (emjburns) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/veto/unhealthy/UnhealthyVeto.kt:48]: You're right that this code will not prevent it, it's different code.

File: keel-core/src/test/kotlin/com/netflix/spinnaker/keel/veto/UnhealthyVetoTests.kt [hunk 21]
```diff
@@ -0,0 +1,146 @@
...
+          }
+        }
+
+        context("resource was marked unhealthy 6 minutes ago") {
+          before {
+            every { unhealthyRepository.getLastALlowedTime(r.id) } returns clock.instant().minusSeconds(60 * 6)
+            check()
+          }
+
+          test("it is vetoed") {
```
[the remark this answers is not in the document]
Comment (emjburns) [reply keel-core/src/test/kotlin/com/netflix/spinnaker/keel/veto/UnhealthyVetoTests.kt:116]: ya, just a case of me copy-pasting and not changing the message. good catch

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/veto/unhealthy/ClusterViewParams.kt (hunk 19 above)
[the remark this answers is not in the document]
Comment (erikmunson) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/veto/unhealthy/ClusterViewParams.kt:10]: yup this is shorthand for query, it maps to whatever populates this substring search field:

Screen Shot 2020-10-01 at 11 49 37 AM

File: keel-echo/src/main/kotlin/com/netflix/spinnaker/keel/echo/ResourceNotifier.kt (hunk 17 above)
[the remark this answers is not in the document]
Comment (luispollo) [reply keel-echo/src/main/kotlin/com/netflix/spinnaker/keel/echo/ResourceNotifier.kt:103]: That probably means I put that field in the wrong place, but oh well. 😄

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/veto/unhealthy/UnhealthyVeto.kt [hunk 22]
```diff
@@ -0,0 +1,118 @@
...
+
+/**
+ * A veto that stops keel from checking a resource while it's unhealthy,
+ * because there's nothing we can do to make it healthy
+ */
+@Component
+class UnhealthyVeto(
+  private val unhealthyVetoRepository: UnhealthyVetoRepository,
+  @Value("\${veto.unhealthy.ignore-duration:PT5M}")
+  private val ignoreDuration: String,
```
[the remark this answers is not in the document]
Comment (emjburns) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/veto/unhealthy/UnhealthyVeto.kt:34]: I'm not sure but I feel like no. I remember trying Duration and having it not work, but i could be wrong

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/veto/unhealthy/UnhealthyVeto.kt [hunk 23]
```diff
@@ -0,0 +1,118 @@
...
+    if (spec !is Monikered) {
+      throw UnsupportedResourceTypeException("Resource kind ${resource.kind} must be monikered to construct resource links")
+    }
+    if (spec !is Locatable<*>) {
+      throw UnsupportedResourceTypeException("Resource kind ${resource.kind} must be locatable to construct resource links")
+    }
+    val locations = spec.locations
+    if (locations !is AccountAwareLocations) {
+      throw UnsupportedResourceTypeException("Resource kind ${resource.kind} must be have account aware locations to construct resource links")
+    }
```
[the remark this answers is not in the document]
Comment (emjburns) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/veto/unhealthy/UnhealthyVeto.kt:83]: No, that is just about artifacts. I need both Monikered and Locatable.

File: keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlNotifierRepository.kt [hunk 24]
```diff
@@ -0,0 +1,77 @@
...
+  override val clock: Clock,
+  private val jooq: DSLContext,
+  private val sqlRetry: SqlRetry
+) : NotifierRepository(clock) {
+  override fun addNotification(resourceId: String, notifier: Notifier): Boolean {
+    sqlRetry.withRetry(READ) {
+      jooq.select(NOTIFIER.NOTIFY_AT)
+        .from(NOTIFIER)
+        .where(NOTIFIER.RESOURCE_ID.eq(resourceId))
+        .and(NOTIFIER.NOTIFIER_.eq(notifier.name))
```
[the remark this answers is not in the document]
Comment (emjburns) [reply keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlNotifierRepository.kt:23]: I'm updating these fields

File: keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlUnhealthyVetoRepository.kt [hunk 25]
```diff
@@ -0,0 +1,103 @@
...
+    }
+  }
+
+  override fun markUnhealthy(resourceId: String, application: String) {
+      sqlRetry.withRetry(RetryCategory.WRITE) {
+        jooq.insertInto(UNHEALTHY_VETO)
+          .set(UNHEALTHY_VETO.RESOURCE_ID, resourceId)
+          .set(UNHEALTHY_VETO.APPLICATION, application)
+          .set(UNHEALTHY_VETO.TIME_DETECTED, clock.millis())
+          .set(UNHEALTHY_VETO.LAST_TIME_ALLOWED, clock.millis())
```
[the remark this answers is not in the document]
Comment (emjburns) [reply keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlUnhealthyVetoRepository.kt:49]: Because we are saying, before this time right now it was allowed to be checked.

File: keel-core/src/test/kotlin/com/netflix/spinnaker/keel/veto/UnhealthyVetoTests.kt [hunk 26]
```diff
@@ -0,0 +1,146 @@
...
+            every { unhealthyRepository.getLastALlowedTime(r.id) } returns clock.instant().minusSeconds(60 * 4)
+            check()
+          }
+
+          test("it is vetoed") {
+            expectThat(result).isNotNull().isNotAllowed()
+          }
+        }
+
+        context("resource was marked unhealthy 6 minutes ago") {
```
[the remark this answers is not in the document]
Comment (emjburns) [reply keel-core/src/test/kotlin/com/netflix/spinnaker/keel/veto/UnhealthyVetoTests.kt:110]: woah i initially put a thumbs down comment here, sorry.

**Comment by emjburns** [comment]
Closing in favor of a different approach
